packages/firestore/src/core/firestore_client.ts

```diff
@@ -391,24 +393,16 @@ export class FirestoreClient {
 
   listen(
     query: Query,
-    observer: Observer<ViewSnapshot>,
-    options: ListenOptions
-  ): QueryListener {
+    options: ListenOptions,
+    observer: AsyncObserver<ViewSnapshot>
+  ): Unsubscribe {
     this.verifyNotTerminated();
     const listener = new QueryListener(query, observer, options);
     this.asyncQueue.enqueueAndForget(() => this.eventMgr.listen(listener));
-    return listener;
-  }
-
-  unlisten(listener: QueryListener): void {
-    // Checks for termination but does not raise error, allowing unlisten after
-    // termination to be a no-op.
-    if (this.clientTerminated) {
-      return;
-    }
-    this.asyncQueue.enqueueAndForget(() => {
-      return this.eventMgr.unlisten(listener);
-    });
+    return () => {
+      observer.mute();
+      this.asyncQueue.enqueueAndForget(() => this.eventMgr.unlisten(listener));
```

Comment:
@schmidt-sebastian Can you just confirm that it is indeed unnecessary to check for `if (this.clientTerminated)` before calling `this.asyncQueue.enqueueAndForget()`? I'm not familiar enough with `AsyncQueue` to know, but the old code avoids calling `this.asyncQueue.enqueueAndForget()` in the case that `this.clientTerminated` is false. The new code will call it regardless of the value of `this.clientTerminated`.

This comment also applies to `addSnapshotsInSyncListener()` below.